library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/process/traversal/strategy/optimisation/GafferPopGraphStepStrategy.java

```diff
@@ -24,28 +24,56 @@
 import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
 import org.apache.tinkerpop.gremlin.process.traversal.step.map.NoOpBarrierStep;
 import org.apache.tinkerpop.gremlin.process.traversal.strategy.AbstractTraversalStrategy;
+import org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.OptionsStrategy;
 import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.opencypher.gremlin.translation.CypherAst;
+import org.opencypher.gremlin.translation.translator.Translator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import uk.gov.gchq.gaffer.tinkerpop.process.traversal.step.GafferPopGraphStep;
 
+import java.util.Optional;
+
 /**
  * The {@link GraphStep} strategy for GafferPop, this will replace the default
  * {@link GraphStep} of a query to add Gaffer optimisations. Such as gathering
  * any {@link HasStep}s so that a Gaffer View can be constructed for the query.
+ * Will also handle the translation any Cypher queries passed in a with() step
+ * into a Gremlin traversal.
  *
  * <pre>
  * g.V().hasLabel()    // replaced by GafferPopGraphStep
  * g.E().hasLabel()    // replaced by GafferPopGraphStep
+ * g.with("cypher", "query") // translated to Gremlin traversal
  * </pre>
  */
 public final class GafferPopGraphStepStrategy extends AbstractTraversalStrategy<TraversalStrategy.ProviderOptimizationStrategy> implements TraversalStrategy.ProviderOptimizationStrategy {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GafferPopGraphStepStrategy.class);
     private static final GafferPopGraphStepStrategy INSTANCE = new GafferPopGraphStepStrategy();
 
+    /**
+     * Key used in a with step to include a opencypher query traversal
+     */
+    public static final String CYPHER_KEY = "cypher";
+
     private GafferPopGraphStepStrategy() {
     }
 
     @Override
     public void apply(final Admin<?, ?> traversal) {
+        // Check for any options on the traversal
+        Optional<OptionsStrategy> optionsStrategy = traversal.getStrategies().getStrategy(OptionsStrategy.class);
+        // Translate and add a cypher traversal in if that key has been set
+        if (optionsStrategy.isPresent() && optionsStrategy.get().getOptions().containsKey(CYPHER_KEY)) {
+            LOGGER.info("Replacing traversal with translated Cypher query");
+            CypherAst ast = CypherAst.parse((String) optionsStrategy.get().getOptions().get(CYPHER_KEY));
```

Comment:
What happens if the Cypher query is malformed/invalid? It doesn't seem that this method throws an exception, at least not a checked one anyway.

Reply:
So you will get a `SyntaxException` and the query will be rejected (gives a reason too) I agree with your other comment probably good to add a test in for it for reference.

Reply:
Any info on this seems to be entirely absent [from their Javadocs](https://www.javadoc.io/doc/org.opencypher.gremlin/translation/1.0.0/index.html) unfortunately. At least the code handles it ok.

Reply:
The javadocs are unfortunate, its probably as the exception is a runtime one so isn't declared by the methods throwing it but there is at least some info about the exception class itself [here](https://www.javadoc.io/static/org.opencypher.gremlin/translation/1.0.0/org/opencypher/gremlin/translation/exception/SyntaxException.html).